brave/brave-core
# Solana auto discovery

Resolves https://github.com/brave/brave-browser/issues/27246
Security review - https://github.com/brave/security/issues/1161

## Summary of changes
Adds support for automatic 'discovery' of Solana assets, meaning when new Solana accounts are added or when the frontend calls DiscoverAssetsOnAllSupportedChains(), we issue [getTokenAccountsByOwner](https://docs.solana.com/developing/clients/jsonrpc-api#gettokenaccountsbyowner) RPC calls to fetch the associated SPL token mint addresses owned by each address, checks if those tokens are in our registry, and adds them as visible asset if so.

This PR integrates this logic in the AssetDiscoveryManager which already has similar logic for Ethereum asset discovery.

I squashed to one commit, so I created this list to guide review:
1. Add `JsonRpcService::GetSolanaTokenAccountsByOwner` methods that implement the [getTokenAccountsByOwner](https://docs.solana.com/developing/clients/jsonrpc-api#gettokenaccountsbyowner) method.  1 RPC request will return all the tokens owned by a single address.
    1. Implement getTokenAccountsByOwner in `components/brave_wallet/browser/solana_requests.cc` to create the request bodies for the RPC call
    1. Modify Rust JSON converter function `convert_uint64_in_object_array_to_string` to support conversions of fields that are not at the top level of the JSON object in the array.
1. Implement core Solana asset discovery logic in AssetDiscoveryManager
    1. Implement `AssetDiscoveryManager::DiscoverSolAssets` which takes a list of Solana account addresses, and calls `JsonRpcService::GetSolanaTokenAccountsByOwner` for each concurrently using a barrier callback
    1. Implement `AssetDiscoveryManager::OnGetSolanaTokenAccountsByOwner` to handle the result of an individual RPC calls by decoding the data (base64, then borsh) to get the mint addresses returned in the RPC call, calls the barrier callback with the decoded list of mint addresses
        1. `AssetDiscoveryManager::DecodeMintAddress` does the borsh decoding to get the address
    1. `AssetDiscoveryManager::MergeDiscoveredSolanaAssets` is called with a list of lists of discovered mint addresses (for for each address).  It dedupes the addresses, then fetches the Sol token Registry
    1. `AssetDiscoveryManager::OnGetSolanaTokenRegistry` cross references the returned mint addresses with the list of token registry, and adds the token if it's in the registry
    1. Calls `AssetDiscoveryManager::CompleteDiscoverAssets` to trigger notifications and/or callbacks for tests
1. Trigger Solana asset discovery when Solana accounts are added
    1. Modify KeyringService such that `NotifyAccountsAdded()` is also called when SOL accounts are added
    1. Add a mojom::CoinType param to `DiscoverAssetsOnAllSupportedChainsAccountsAdded`, and if it's SOL, call `DiscoverSolAssets`
1. Trigger Solana asset discovery in `BraveWalletService::DiscoverAssetsOnAllSupportedChains` (this is what's called by frontends)
    1. Modify `BraveWalletService::DiscoverAssetsOnAllSupportedChains` such that it also fetches the users' Solana addresses to perform asset discovery with (calls `AssetDiscoveryManager::DiscoverAssetsOnAllSupportedChainsRefresh`)
    1. Modify `AssetDiscoveryManager::DiscoverAssetsOnAllSupportedChainsRefresh` such that it takes a map of mojom::CoinType -> vector of addresses, representing the addresses for each CoinType the user has, and now call `DiscoverSolAssets` as well as `DiscoverEthAssets` (eth)

## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
Set up an account on Solana mainnet that has tokens listed our token registry, but are not added as visible assets. Add this account using each of these methods, and verify the asset is now a visible asset:
* Create wallet
* Restore wallet
* Add account (standard way)
* Import private key
* Import hardware account

When frontends have implemented calling the `DiscoverAssetsOnAllSupportedChains()` upon page refresh, then Solana tokens should automatically be added because of the work in this PR, but I don't believe there's a way to test that now, so that should be tested when those changes land.
